Add file limit to cpio file creation

andrewsun2898 opened the pull request:
There exists a race condition when attempting to add records to the cpio and when the underlying file is being written to at the same time. Introducing a limit writer in case more data gets written to the file.

andrewsun2898 commented:
Note: Tested this change by adding in a print statement right before record creation with log recording enabled. This almost certainly guaranteed that the file would be written to at the same time the record was being created.

hugelgupf commented:
My original change had a lazy_test.go - can you add that?

On Tue, Apr 18, 2023, 15:50 andrewsun2898 ***@***.***> wrote:

> Note: Tested this change by adding in a print statement right before
> record creation with log recording enabled. This almost certainly
> guaranteed that the file would be written to at the same time the record
> was being created.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/u-root/u-root/pull/2659#issuecomment-1513881249>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPG3EXOBSJAYCDU64LM22DXB4LDHANCNFSM6AAAAAAXDIJ5BE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

andrewsun2898 commented:
> My original change had a lazy_test.go - can you add that?

Sorry! Added that in. I've also added a retry mechanism for the option to retry if it fails specifically on the write with the intention that if it does fail we can catch as much of the file as possible.

rminnich reviewed (commented):
being written to by who? I find the description confusing. A bit more explanation would be useful, for future readers of the commit. Under what scenario does this problem occur? Can you show a command line example?

andrewsun2898 commented:
> being written to by who? I find the description confusing. A bit more explanation would be useful, for future readers of the commit. Under what scenario does this problem occur? Can you show a command line example?

Added a sentence for the specific case I'm running into. Let me know if it's still confusing to you.

rminnich reviewed (commented):
hi, sorry to be so confused, what is the specific command that you typed that leads to this problem, I'm still a bit lost. If I do understand the problem, I suspect this may not be quite the way to solve it.

andrewsun2898 commented:
> hi, sorry to be so confused, what is the specific command that you typed that leads to this problem, I'm still a bit lost. If I do understand the problem, I suspect this may not be quite the way to solve it.

Sorry I have no specific command, but it's a race condition that appears when we enable logging. When we tee log.Print output to a file, there's a chance that when we try and package the log file into a cpio it will fail because of some stray log.Print statement. An easy way to trigger this race condition is to enable logging and add a print statement just before cpio creation. 

This change prevents the file from being written past the limit which will be the file size before the cpio creation.

rminnich commented:
OK, I finally get it. Sorry to be so dense. I think I would phrase this as follows:

"If cpio is asked to archive a file, and that file is growing, the cpio may never complete. This change will cause cpio to stat the files as they are traversed, and only write as many bytes of the file as it has at that time. This change ensures, further, that should some of the files be named IPC endpoints (as can happen on, e.g., Plan 9), they will stat as zero bytes, limiting any problematical side effects for both cpio and the IPC endpoint.

This problem has many names, but one name readers may recognize is 'phase error' from old school tar.
"

something like that. This is a very good PR and I'm sorry I dragged it on so long, but the description confused me (and a few others).

andrewsun2898 commented:
No worries, it's my bad for not being able to explain clearer. I liked your description so I added to the commit message. Thanks for reviewing!

andrewsun2898 commented:
As it's implemented currently in u-root, cpio will error out which is why I wrote it like that. Personally I think it's ok since it describes what u-root currently does when you encounter this situation.
